File: .Trash-1000/files/simple-func.py
#!/usr/bin/python3
# simple-func.py — eBPF 函数调用实验加载器
# 功能：加载 simple-func.c，监控 execve 事件，打印 PID 和进程名
# 用法：sudo python3 simple-func.py

from bcc import BPF

b = BPF(src_file="simple-func.c")

def print_event(cpu, data, size):
    event = b["events"].event(data)
    print(f"PID={event.pid:6d} COMM={event.comm.decode():16s}")

b["events"].open_perf_buffer(print_event)

print("实验 1: eBPF 函数调用实验")
print("监控中... 在另一个终端执行命令观察输出")
print("按 Ctrl+C 退出\n")

try:
    while True:
        b.perf_buffer_poll()
except KeyboardInterrupt:
    print("\n退出。eBPF 程序已随进程自动卸载。")
    print("\n💡 如果想用 bpftool 观察字节码，请在程序运行时另开终端执行：")
    print("   sudo bpftool prog list | grep execve")
    print("   sudo bpftool prog dump xlated id <ID>")

File: .Trash-1000/files/simple-func.c
// simple-func.c — eBPF 中的函数调用实验
// 功能：把"获取进程信息"提取为辅助函数，演示 BPF-to-BPF 调用
// 用法：sudo python3 simple-func.py

#include <uapi/linux/ptrace.h>
#include <linux/sched.h>

struct data_t {
    u32 pid;
    char comm[16];
};

BPF_PERF_OUTPUT(events);

// 🔥 辅助函数：填充进程信息
// __always_inline 强制内联展开，避免函数调用栈开销
static __always_inline void fill_proc_info(struct data_t *data) {
    data->pid = bpf_get_current_pid_tgid() >> 32;
    bpf_get_current_comm(&data->comm, sizeof(data->comm));
}

TRACEPOINT_PROBE(syscalls, sys_enter_execve) {
    struct data_t data = {};

    // 调用辅助函数
    fill_proc_info(&data);

    events.perf_submit(args, &data, sizeof(data));
    return 0;
}
